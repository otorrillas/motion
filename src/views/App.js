import React, { PureComponent, Suspense } from 'react';

import styles from './App.module.css';

import { fetchTrackList } from '../api';
import { PlayerContext } from './PlayerContext';

import Loader from '../components/Loader';
import TrackList from '../components/TrackList';
import Player from '../components/Player';

class App extends PureComponent {
  state = {
    trackList: null,
    currentTrack: null,
    isPlaying: false,
    playerRef: React.createRef()
  };

  componentDidMount() {
    this._loadAsyncData();
  }

  componentWillUnmount() {
    if (this._asyncRequest) {
      this._asyncRequest.cancel();
    }
  }

  _loadAsyncData() {
    this._asyncRequest = fetchTrackList().then(trackList => {
      this._asyncRequest = null;
      this.setState({ trackList });
    });
  }

  onTrackSelect = track => {
    const { currentTrack, isPlaying } = this.state;

    if (currentTrack && track === currentTrack) {
      const action = isPlaying ? this.pause : this.play;
      action();
    } else {
      this.setState({
        currentTrack: track,
        isPlaying: true
      });
    }
  };

  pause = () => {
    this.state.playerRef.current.pause();
    this.setState({
      isPlaying: false
    });
  };

  play = () => {
    this.state.playerRef.current.play();
    this.setState({
      isPlaying: true
    });
  };

  render() {
    const { currentTrack, trackList, playerRef } = this.state;

    return (
      <Suspense fallback={<Loader />} maxDuration={2000}>
        <div className={styles.app}>
          <h1>Motion</h1>

          <PlayerContext.Provider value={this.state}>
            <TrackList songs={trackList} onSelect={this.onTrackSelect} />

            <Player track={currentTrack} ref={playerRef} />
          </PlayerContext.Provider>
        </div>
      </Suspense>
    );
  }
}

export default App;
