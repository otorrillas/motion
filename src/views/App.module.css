.app {
  color: white;
  text-align: center;

  padding-top: 4rem;
  /* display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center; */
}
