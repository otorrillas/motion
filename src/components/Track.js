import React from 'react';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlay, faPause } from '@fortawesome/free-solid-svg-icons';

import styles from './Track.module.css';
import cx from 'classnames';
import { PlayerContext } from '../views/PlayerContext';

const Track = ({ track, onClick }) => (
  <PlayerContext.Consumer>
    {({ currentTrack, isPlaying }) => {
      const { id, title, artist, album } = track;

      const isSelected = currentTrack && currentTrack.id === id;
      const icon = isSelected && isPlaying ? faPause : faPlay;

      return (
        <tr
          className={cx(styles.track, { [styles.selected]: isSelected })}
          onClick={onClick}
        >
          <td>
            <FontAwesomeIcon icon={icon} />
          </td>
          <td>{title}</td>
          <td>{artist}</td>
          <td>{album}</td>
        </tr>
      );
    }}
  </PlayerContext.Consumer>
);

export default Track;
