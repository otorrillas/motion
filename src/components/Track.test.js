import React from 'react';
import renderer from 'react-test-renderer';

import Track from './Track';

import { PlayerContext } from '../views/PlayerContext';

describe('<Track />', () => {
  const mockActions = {
    onClick: jest.fn()
  };

  const requiredProps = {
    track: {
      id: 123,
      title: 'sample title',
      artist: 'sample artist',
      album: 'sample album'
    }
  };

  const renderComponent = (
    value = { currentTrack: null, isPlaying: false },
    props
  ) =>
    renderer.create(
      <PlayerContext.Provider value={value}>
        <Track {...requiredProps} {...mockActions} {...props} />
      </PlayerContext.Provider>
    );

  describe('@render', () => {
    it('default', () => {
      const tree = renderComponent();
      expect(tree).toMatchSnapshot();
    });

    it('with current track matching & playing', () => {
      const tree = renderComponent({
        currentTrack: requiredProps.track,
        isPlaying: true
      });
      expect(tree).toMatchSnapshot();
    });
  });

  describe('@actions', () => {
    beforeEach(() => {
      jest.restoreAllMocks();
    });
    describe('onClick', () => {
      it('click on track calls onClick', () => {
        const component = renderComponent();
        component.root.findByType('tr').props.onClick();

        expect(mockActions.onClick).toHaveBeenCalled();
      });
    });
  });
});
