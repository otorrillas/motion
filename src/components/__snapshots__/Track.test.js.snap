// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Track /> @render default 1`] = `
<tr
  className="track"
  onClick={[MockFunction]}
>
  <td>
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-play fa-w-14 "
      data-icon="play"
      data-prefix="fas"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </td>
  <td>
    sample title
  </td>
  <td>
    sample artist
  </td>
  <td>
    sample album
  </td>
</tr>
`;

exports[`<Track /> @render with current track matching & playing 1`] = `
<tr
  className="track selected"
  onClick={[MockFunction]}
>
  <td>
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-pause fa-w-14 "
      data-icon="pause"
      data-prefix="fas"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M144 479H48c-26.5 0-48-21.5-48-48V79c0-26.5 21.5-48 48-48h96c26.5 0 48 21.5 48 48v352c0 26.5-21.5 48-48 48zm304-48V79c0-26.5-21.5-48-48-48h-96c-26.5 0-48 21.5-48 48v352c0 26.5 21.5 48 48 48h96c26.5 0 48-21.5 48-48z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </td>
  <td>
    sample title
  </td>
  <td>
    sample artist
  </td>
  <td>
    sample album
  </td>
</tr>
`;
